Dlg string handling

shadrackkibet opened the pull request:
@africanmathsinitiative/developers this is in progress

Lunalo commented:
@africanmathsinitiative/developers 

Someone should test this.

It is ready

lilyclements commented:
@shadrackkibet I'm about to review this. However, I've noticed you've made changes to the Main Form. Do try to avoid making changes to frmMain in the future on a branch containing other dialog work. This is to avoid conflicts.

shadrackkibet commented:
@lilyclements I noticed that and i have reverted the changes i made on frmMain

stevenndungu commented:
@lilyclements  Was reviewing it, in case you have additional comments then that will be useful as well

lilyclements commented on instat/dlgStringHandling.Designer.vb:
```diff
@@ -205,11 +151,65 @@ Partial Class dlgStringHandling
+        Me.ucrPnlStringHandling.Size = New System.Drawing.Size(294, 51)
         Me.ucrPnlStringHandling.TabIndex = 16
         '
         'dlgStringHandling
```
General Comments on the layout:

Looks okay, lots of things need to be tidied up here:
* Move the buttons at the top around so they are all the same size. Some are overlapping others (e.g. Count) to make them look smaller
* lblRepalceBy is not right. You have spelt Replace wrong so need to correct this
* The tab order needs to be addressed; ask someone at the office to help you with this
* You need to align the "lefts" of your labels and receivers. I sent out a document about this a while ago. Have a look at that, or ask someone in the office for a copy.
* ucrSave and ucrBase are too low down and very spread apart from one another, these can be brought closer together.
* Your labels are at different distances from your receivers for each one. Have a look at the document about where these should be placed
* Your first receiver is too high up compared to the ucrSelector. You want the top of the receiver to align with the top of the listbox in the ucrSelector
* You need a space in your text in lblRepalceBy

lilyclements commented on instat/dlgStringHandling.vb:
```diff
@@ -8,7 +8,7 @@
 ' This program is distributed in the hope that it will be useful,
 ' but WITHOUT ANY WARRANTY; without even the implied warranty of
-' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+' MERCHANTABILITY or FITNESS FOR A URPOSE.  See the
```
I assume this was an accident. However, it needs to be changed back

lilyclements commented on instat/dlgStringHandling.vb:
```diff
@@ -33,27 +34,123 @@ Public Class dlgStringHandling
+        ucrInputPattern.SetParameter(New RParameter("pattern", 1))
+        ucrInputReplaceBy.SetParameter(New RParameter("replacement", 2))
+
+
```
Avoid having two blank lines between code. You have done this elsewhere in the dialog so look out for this

lilyclements commented on instat/dlgStringHandling.vb:
```diff
@@ -33,27 +34,123 @@ Public Class dlgStringHandling
+        ucrInputReplaceBy.SetRCode(clsReplaceFunction, bReset)
+
+    End Sub
+    Private Sub ChangeBaseFunction()
```
Between the end of a sub and start of a sub, there should be a space

lilyclements commented on instat/dlgStringHandling.vb:
```diff
@@ -33,27 +34,123 @@ Public Class dlgStringHandling
+    Private Sub ChangeBaseFunction()
+        If rdoCount.Checked Then
+            ucrBase.clsRsyntax.SetBaseRFunction(clsCountFunction)
+            ucrSave.SetPrefix("Count")
```
Either all prefixes should have a capital first letter, or none of them should. Consistency would be good here.
I don't mind which - @dannyparsons do you remember which one we agreed on?

lilyclements commented on instat/dlgStringHandling.vb:
```diff
@@ -33,27 +34,123 @@ Public Class dlgStringHandling
+        ucrSave.SetSaveTypeAsColumn()
+        ucrSave.SetDataFrameSelector(ucrSelectorStringHandling.ucrAvailableDataFrames)
+        ucrSave.SetIsComboBox()
+        ucrSave.SetLabelText("Save Result to:")
```
This can just read "Save Result:" for consistency with other dialogs

lilyclements commented on instat/dlgStringHandling.vb:
```diff
@@ -33,27 +34,123 @@ Public Class dlgStringHandling
+            ucrSave.SetPrefix("extract")
+            ucrBase.clsRsyntax.SetBaseRFunction(clsExtractFunction)
+        ElseIf rdoLocate.Checked Then
+            ucrSave.SetPrefix("locate")
```
Locate makes several columns at once, this means that the assign to for locate should be different and more like how it is in dlgSplitText.
This is because currently my default column names for Locate are Locate11 and Locate12 as opposed to Locate1 and Locate2

shadrackkibet replied:
My understanding is that the locate function should append two columns at the same time and that is why they are Locate11 and Locate12 which gives the "start" and the "end" of a pattern .I think this is okay according to it's functionality in R.

lilyclements replied:
So, when we have a ucrSave that produces multiple columns, the code is different to usual. For example, in dlgSplitText. You want this prefix to be added internally to the name so that by one click of the button you get Locate1, Locate2. This means, I believe, a few minor changes. Have a look at dlgSplitText and see what is different on the save control and the assign to control

lilyclements commented:
@shadrackkibet your changes on the main form were not reverted by re-disabling the dialog.
@Lunalo could you show @shadrackkibet how to revert a commit on Visual Studio for future reference for him?

shadrackkibet commented:
@stevenndungu and @lilyclements i have made the necessary changes.

lilyclements commented:
Once you've merged my changes I made in, there are still some other things to address. For example, I might change the new name in the ucrSave, then change the button above. If I do this, then I lose my new name I inputted in ucrSave.
See dlgRenameSheet or dlgName for when the user changes the name and then changes the sheet or variable on how to tackle this.

Additionally, `.SetName` for all the functions is not the way to fix the problem. You want to have .SetPrefix for most of them, and need to sort out the case of the "locate" button separately. It isn't the same code for all buttons

Finally, I just noticed that when "locate" is selected, there is a capital "L" now on the default name. Either all default names should have capitals or they all should not

shadrackkibet commented:
@dannyparsons . you can check on this.

stevekogo commented:
@shadrackkibet  You invited @dannypasrsons to look at this, has all the comments from the reviewers been addressed?

shadrackkibet commented:
I have addressed most of the comments suggested by reviewers except transforming this to the new system.@dannyparsons was to merge this as it is in the old system and I am hoping this to happen today. i will be able to make the suggested changes and hope this will be ready for Re-Review soon.

dannyparsons commented:
I'll merge this now as discussed so it can be updated.
